internal/models/flow_state.go

```diff
@@ -0,0 +1,89 @@
+package models
+
+import (
+	"database/sql"
+	"github.com/pkg/errors"
+	"github.com/supabase/gotrue/internal/storage"
+	"time"
+
+	"github.com/gofrs/uuid"
+)
+
+type FlowState struct {
+	ID                   uuid.UUID `json:"id" db:"id"`
+	UserID               uuid.UUID `json:"user_id" db:"user_id"`
+	AuthCode             string    `json:"auth_code" db:"auth_code"`
+	CodeChallenge        string    `json:"code_challenge" db:"code_challenge"`
+	CodeChallengeMethod  string    `json:"code_challenge_method" db:"code_challenge_method"`
+	ProviderType         string    `json:"provider_type" db:"provider_type"`
+	ProviderAccessToken  string    `json:"provider_access_token" db:"provider_access_token"`
+	ProviderRefreshToken string    `json:"provider_refresh_token" db:"provider_refresh_token"`
+	CreatedAt            time.Time `json:"created_at" db:"created_at"`
+	UpdatedAt            time.Time `json:"updated_at" db:"updated_at"`
+}
+
+type CodeChallengeMethod int
+
+const (
+	SHA256 CodeChallengeMethod = iota
+	Plain
+)
+
+func (authMethod CodeChallengeMethod) String() string {
+	switch authMethod {
+	case SHA256:
+		return "s256"
+	case Plain:
+		return "plain"
+	}
+	return ""
+}
+
+func (FlowState) TableName() string {
+	tableName := "flow_state"
+	return tableName
+}
+
+func NewFlowState(providerType, codeChallenge string, codeChallengeMethod CodeChallengeMethod) (*FlowState, error) {
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, errors.New("error generating unique oauth state verifier")
+	}
+	authCode, err := uuid.NewV4()
+	if err != nil {
+		return nil, errors.New("error generating auth code")
+	}
```

Comment:
Just recently learned you can use `uuid.Must(uuid.NewV4())` -- it is practically impossible to not get a UUID. If the random number generator of the OS is broken, there's a larger problem and GoTrue won't be able to even start.